girder/utility/plugin_utilities.py

```diff
@@ -189,40 +220,42 @@ def findAllPlugins(curConfig=None):
               'loaded.'))
         return allPlugins
-    dirs = [dir for dir in os.listdir(pluginsDir) if os.path.isdir(
-            os.path.join(pluginsDir, dir))]
-
-    for plugin in dirs:
-        data = {}
-        configJson = os.path.join(pluginsDir, plugin, 'plugin.json')
-        configYml = os.path.join(pluginsDir, plugin, 'plugin.yml')
-        if os.path.isfile(configJson):
-            with open(configJson) as conf:
-                try:
-                    data = json.load(conf)
-                except ValueError as e:
-                    print(TerminalColor.error(
-                        'ERROR: Plugin "%s": plugin.json is not valid JSON.' %
-                        plugin))
-                    print(e)
-                    continue
-        elif os.path.isfile(configYml):
-            with open(configYml) as conf:
-                try:
-                    data = yaml.safe_load(conf)
-                except yaml.YAMLError as e:
-                    print(TerminalColor.error(
-                        'ERROR: Plugin "%s": plugin.yml is not valid YAML.' %
-                        plugin))
-                    print(e)
-                    continue
-
-        allPlugins[plugin] = {
-            'name': data.get('name', plugin),
-            'description': data.get('description', ''),
-            'version': data.get('version', ''),
-            'dependencies': set(data.get('dependencies', []))
-        }
+
+    for pluginDir in pluginsDir:
+        dirs = [dir for dir in os.listdir(pluginDir) if os.path.isdir(
+        os.path.join(pluginDir, dir))]
+
+        for plugin in dirs:
+            data = {}
+            configJson = os.path.join(pluginDir, plugin, 'plugin.json')
+            configYml = os.path.join(pluginDir, plugin, 'plugin.yml')
+            if os.path.isfile(configJson):
+                with open(configJson) as conf:
+                    try:
+                        data = json.load(conf)
+                    except ValueError as e:
+                        print(TerminalColor.error(
+                            'ERROR: Plugin "%s": plugin.json is not valid JSON.' %
+                            plugin))
+                        print(e)
+                        continue
+            elif os.path.isfile(configYml):
+                with open(configYml) as conf:
+                    try:
+                        data = yaml.safe_load(conf)
+                    except yaml.YAMLError as e:
+                        print(TerminalColor.error(
+                            'ERROR: Plugin "%s": plugin.yml is not valid YAML.' %
+                            plugin))
+                        print(e)
+                        continue
+
+            allPlugins[plugin] = {
+                'name': data.get('name', plugin),
+                'description': data.get('description', ''),
+                'version': data.get('version', ''),
+                'dependencies': set(data.get('dependencies', []))
+            }
 
```

Comment:
I think @cjh1 and @jbeezley are discussing 2 separate issues, let me know if I made any wrong assumptions.

@jbeezley is simply saying that the entire function call should be memoized because it hits the filesystem a lot. I agree, seeing as how it's called in other places and even part of an API call. The only issue that arises here is ```curConfig``` is a dict, and so caching the results given certain arguments is a little bit more difficult (not sure if we want to start dealing with frozensets and so on). We could always cache it for the ```None``` case which I assume is dominant based on the rest of core's code.

@cjh1 Correct me if I'm wrong, but I think you're saying ```allPlugins[plugin]``` should contain a ```pluginParentDir``` key? If that's the case, I'm simply saying that everywhere ```allPlugins``` is used, it's handled by referring to the name of the plugin, which means it wouldn't yield any benefit for ```loadPlugin``` or anything else. That *could* be beneficial if we reworked ```loadPlugin``` to be able to handle the plugin dictionary, which might be something that @zachmullen considered?